home/tests.py

```diff
@@ -208,3 +212,171 @@ def tags_response(self, client):
+            # User entered all the required fields and sub-subject and book
+            ({'title': "Question in Math", 'content': 'How much is it 1+1?',
+              'subject': 1, 'grade': Grade.GRADE7, 'book_page': 23}),
+            # User entered all the required fields and a valid book page
+        ])
+        @pytest.mark.django_db
+        def test_post_valid_question_with_client(self, client, valid_data):
+            with pytest.raises(Question.DoesNotExist):
+                assert Question.objects.get(title=valid_data["title"])
+
+            client.login(username='Lior', password='LiorLior')
+            client.post('/explore/new_question', data=valid_data)
+
+            assert Question.objects.filter(title=valid_data["title"]).exists()
+
+        @pytest.mark.parametrize("invalid_data", [
+            ({'title': "Question in Math", 'subject': 1, 'grade': Grade.GRADE7}),
+            # User entered all the required fields except content
+            ({'title': "Question in Math", 'content': 'How much is it 1+1?', 'grade': Grade.GRADE7}),
+            # User entered all the required fields except subject
+            ({'title': "Question in Math", 'content': 'How much is it 1+1?', 'subject': 1}),
+            # User entered all the required fields except grade
+            ({'title': "Question in Math", 'content': 'How much is it 1+1?', 'subject': 1, 'book_page': -23}),
+            # User entered all the required fields and a negative book page
+            ({'title': "Question in Math", 'content': 'How much is it 1+1?', 'subject': 1, 'book_page': 43343}),
+            # User entered all the required fields and to high book page
+        ])
+        @pytest.mark.django_db
+        def test_post_invalid_question_with_client(self, client, invalid_data):
+            client.login(username='Lior', password='LiorLior')
+            client.post('/explore/new_question', data=invalid_data)
+
+            with pytest.raises(Question.DoesNotExist):
+                assert Question.objects.get(title=invalid_data["title"])
+
+        @pytest.mark.django_db
+        def test_post_question_without_title_with_client(self, client):
+            client.login(username='Lior', password='LiorLior')
+            client.post('/explore/new_question', data={'content': 'How much is it 1+1?',
+                                                       'subject': 1, 'grade': Grade.GRADE7})
+
+            with pytest.raises(Question.DoesNotExist):
+                assert Question.objects.get(content='How much is it 1+1?')
+
+        @pytest.mark.django_db
+        def test_question_form_and_template_displayed(self, client):
+            response = client.get('/explore/new_question')
+
+            assert response.status_code == 200
+            assert isinstance(response.context['form'], type(QuestionForm))
+            assertTemplateUsed(response, 'home/questions/new_question.html')
+
+
+class TestProfileMiddleware():
```

Comment:
This is a a somewhat fragile test, since it rlies too heavily on the internal implementation of the middleware. It is also not verifying the the correct functionality is achieved from the point of view of a view function.

The way I would test it would be:
1. Use `monkeypatch` to replace a view function with a mockup:
2. Perform requests using the `client` fixture and its [`force_login` and `logout` methods](https://docs.djangoproject.com/en/3.2/topics/testing/tools/#django.test.Client.force_login)
3. Inspect the `request` structure that was passed to the view mockup to assert the existance and the contents of the `profile` attribute.

Given that the test does partially work, you may decide not to implement this at this point.

Reply:
At the moment I prefer to leave the implementation as it is, as long as it is acceptable enough, and later @liornoy and I together will improve the test in the way you mentioned.
I opened an issue for that.